### PR TITLE
mount - add a newline at the end of line in fstab

### LINKED ISSUE #210: mount module does not add mountpoint correctly if /etc/fstab does not end with a newline
##### SUMMARY
mount module does not add mountpoint correctly if /etc/fstab does not end with a newline.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
- ansible.posix.mount

##### ANSIBLE VERSION
```
2.12.0.dev0
```

##### CONFIGURATION
[1] ansible.cfg
```
$ ansible-config dump --only-changed
CALLBACKS_ENABLED(/home/hsaito/work/testing/collections/ansible.cfg) = ['ansible.posix.profile_tasks']
COLLECTIONS_PATHS(/home/hsaito/work/testing/collections/ansible.cfg) = ['/home/hsaito/work/testing/collections']
DISPLAY_SKIPPED_HOSTS(/home/hsaito/work/testing/collections/ansible.cfg) = False
HOST_KEY_CHECKING(/home/hsaito/work/testing/collections/ansible.cfg) = False
```
[2] Playbook
```
---
- hosts: linux
  gather_facts: false

  tasks:
    - ansible.posix.mount:
        src: /dev/vdb1
        path: /mnt
        opts: defaults
        state: mounted
        fstype: xfs
        state: mounted
      become: true
```
[3] There is no newline at the end of file of /etc/fstab:
```
[root@server00 ~]# cat /etc/fstab 
... snip ...
/dev/mapper/cl-root     /                       xfs     defaults        0 0
UUID=fd99bb22-3376-434f-b976-f0e0845ddf1b /boot                   ext4    defaults        1 2
/dev/mapper/cl-swap     swap                    swap    defaults        0 0

# No newline at EOF[root@server00 ~]# 

[root@server00 ~]# od -xc /etc/fstab 
... snip ...
0001100    3020    0a0a    2023    6f4e    6e20    7765    696c    656e
              0  \n  \n   #       N   o       n   e   w   l   i   n   e
0001120    6120    2074    4f45    0046
              a   t       E   O   F
0001127
```

##### OS / ENVIRONMENT
- Control Node: Fedora34
- Managed Node: CentOS Stream release 8

##### STEPS TO REPRODUCE

Launch playbook [2] to manage /etc/fstab [3]

```
$ ansible-playbook -i inventory playbook/mount/mount_linux_xfs.yml 

PLAY [linux] *********************************************************************************************************
Tuesday 15 June 2021  22:03:54 +0900 (0:00:00.028)       0:00:00.028 ********** 

TASK [ansible.posix.mount] *******************************************************************************************
changed: [server00]

PLAY RECAP ***********************************************************************************************************
server00                   : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

Tuesday 15 June 2021  22:03:55 +0900 (0:00:00.535)       0:00:00.564 ********** 
=============================================================================== 
ansible.posix.mount ------------------------------------------------------------------------------------------- 0.54s
```

##### EXPECTED RESULTS
mount succeeded and new mountpoint added to /etc/fstab correctly:
```
[root@server00 ~]# mount
... snip ...
/dev/vdb1 on /mnt type xfs (rw,relatime,attr2,inode64,noquota)

[root@server00 ~]# cat /etc/fstab 
... snip ...
# No newline at EOF
/dev/vdb1 /mnt xfs defaults 0 0
```

##### ACTUAL RESULTS
mount succeeded, but incorrect mountpoint was added to /etc/fstab:
```
... snip ...
# No newline at EOF/dev/vdb1 /mnt xfs defaults 0 0
```

### PR DESCRIPTION
**Pull request by saito-hideki**
##### SUMMARY

Modify mount module to add a newline at the end of line in `/etc/fstab`.

Fixes #210

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
- ansible.posix.mount

##### ADDITIONAL INFORMATION
None

### CODE REVIEW

File: plugins/modules/mount.py [hunk 1]
```diff
@@ -247,6 +247,10 @@ def _set_mount_save_old(module, args):
             '%(src)s - %(name)s %(fstype)s %(passno)s %(boot)s %(opts)s\n')
 
     for line in open(args['fstab'], 'r').readlines():
+        # Append newline if the line in fstab does not finished with newline.
```
Comment (Akasurde) [inline plugins/modules/mount.py:250]: @saito-hideki Thanks for the PR. Would it possible to add a test for this change? Thanks.

Comment (saito-hideki) [reply plugins/modules/mount.py:250]: @Akasurde sure thing! I'll add integration tests

Comment (saito-hideki) [reply plugins/modules/mount.py:250]: done! :)

**Comment by saito-hideki** [comment]
added integration tests to this PR.

**Review by quidame — APPROVED** [review]
@saito-hideki thanks for fixing this.

It's really a strange thing to get missing EOL in fstab. First, POSIX has stated that a file MUST end with a EOL (so every lines in the file end with a newline, that makes them all parsable the same way). And (afaik) there is no standard tool/command to handle this file, so my opinion is that some custom script introduced this no-eol.

Even if it's out of the scope of this PR, what leads to deposixify a standard configuration file is still questionable :)

**Comment by saito-hideki** [comment]
Hi @quidame, Thank you for always giving me valuable advice!
Yeah, I think so too,  according to the POSIX standard, the EOL of text files should be ended a newline.

BTW, I have been trying to replicate issue #180(but I can't replicate it yet). In the middle of the investigation for the issue, I saw this behavior without a newline in the fstab file.
So, I'm thinking that if the cause of this issue #180 is regarding the custom user script like you pointed out, this PR would possibly prevent those problems :)

Thanks!

**Comment by quidame** [comment]
@BartokIT, is it possible for you to test this version, to ensure it covers usecase described in #180 ?

**Comment by saito-hideki** [comment]
Rebased to the current main branch.

**Pull request merged**